cli/replicasets/setup.go

```diff
@@ -0,0 +1,287 @@
+		}
+	}
+
+	consoleSockPath := project.GetInstanceConsoleSock(ctx, controlInstanceName)
+	conn, err := common.ConnectToTarantoolSocket(consoleSockPath)
+	if err != nil {
+		return fmt.Errorf("Failed to connect to Tarantool instance: %s", err)
+	}
+
+	log.Debugf("Connected to %s", consoleSockPath)
+
+	topologyReplicasets, err := getTopologyReplicasets(conn)
+	if err != nil {
+		return fmt.Errorf("Failed to get current topology replicasets: %s", err)
+	}
+
+	var errors []error
+
+	log.Debugf("Setup replicasets")
+
+	for _, replicasetConf := range *replicasetsList {
+		res := common.Result{
+			ID: replicasetConf.Alias,
+		}
+
+		topologyReplicaset := topologyReplicasets.GetByAlias(replicasetConf.Alias)
+
+		if topologyReplicaset == nil {
+			if err := createReplicaset(conn, replicasetConf, instancesConf); err != nil {
+				res.Status = common.ResStatusFailed
+				res.Error = err
+			} else {
+				res.Status = common.ResStatusCreated
+			}
+		} else {
+			if changed, err := updateReplicaset(conn, topologyReplicaset, replicasetConf, instancesConf); err != nil {
+				res.Status = common.ResStatusFailed
+				res.Error = err
+			} else if changed {
+				res.Status = common.ResStatusUpdated
+			} else {
+				res.Status = common.ResStatusOk
+			}
+		}
+
+		if res.Status == common.ResStatusFailed {
+			errors = append(errors, res.FormatError())
+		}
+
+		log.Infof(res.String())
+	}
+
+	if len(errors) > 0 {
+		for _, err := range errors {
+			log.Errorf("%s", err)
+		}
+		return fmt.Errorf("Failed to setup replicasets")
+	}
+
+	if ctx.Replicasets.BootstrapVshard {
+		if err := bootstrapVshard(conn); err != nil {
+			return fmt.Errorf("Failed to bootstrap vshard: %s", err)
+		}
+
+		log.Infof("Vshard is bootstrapped successfully")
+	}
+
+	log.Infof("Topology is set up successfully")
+
+	return nil
+}
+
+func getReplicasetsList(ctx *context.Ctx) (*ReplicasetsList, error) {
+	var err error
+
+	if _, err := os.Stat(ctx.Replicasets.File); err != nil {
+		return nil, fmt.Errorf("Failed to use replicasets configuration file: %s", err)
+	}
+
+	fileContentBytes, err := common.GetFileContentBytes(ctx.Replicasets.File)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read replicasets configuration file: %s", err)
+	}
+
+	var replicasetsConf ReplicasetsConf
+	if err := yaml.Unmarshal([]byte(fileContentBytes), &replicasetsConf); err != nil {
+		return nil, fmt.Errorf("Failed to parse replicasets configuration file %s: %s", ctx.Replicasets.File, err)
+	}
+
+	replicasetsList := make(ReplicasetsList, len(replicasetsConf))
+
+	i := 0
+	for replicasetAlias, replicasetConf := range replicasetsConf {
+		replicasetConf.Alias = replicasetAlias
+
+		replicasetsList[i] = replicasetConf
+		i++
+	}
+
+	// move vshard-router replicasets at the list beginning
+	// to prevent "no remotes with role vshard-router avaliable" error
+	// that happens when vshard-router replicaset is set up last
+	// Probably, this relate to https://github.com/tarantool/cartridge/issues/1139
```

Comment:
Even if it's true it is already fixed: https://github.com/tarantool/cartridge/issues/1139. However, I don't think it's related.

Reply:
Yes, but this version of cartridge isn't released yet. Even if it was released, what about users who didn't update cartridge right after release?

Reply:
Currently I suggest to use some timeout for that.